File: core/src/com/intbyte/bw/engine/entity/Drop.java
package com.intbyte.bw.engine.entity;

import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.intbyte.bw.engine.graphic.Graphic;
import com.intbyte.bw.engine.item.Container;
import com.intbyte.bw.engine.physic.PhysicData;
import com.intbyte.bw.engine.block.Block;

public class Drop extends Entity {

    protected ModelInstance modelInstance;
    protected DropData dropData;


    public Drop(int blockID, DropData dropData) {
        this.rotate = (float) Math.random() * 360;
        this.modelInstance = Block.getBlock(blockID).getModelInstance();
        this.dropData = dropData;
    }

    public Drop(ModelInstance modelInstance, DropData dropData){
        this.rotate = (float) Math.random() * 360;
        this.modelInstance = modelInstance;
        this.dropData = dropData;
    }

    @Override
    protected void spawn() {
        PhysicData data = (PhysicData) getBody().getUserData();
        data.setObject(this);
        data.setType(data.DROP);
    }

    @Override
    public ModelInstance getEntityModel() {
        return modelInstance;
    }

    @Override
    public void render() {
        calculateModelPositionAndRotation(0, 1f, 0);
        float scaleX = modelInstance.transform.getScaleX(),
              scaleY = modelInstance.transform.getScaleY(),
              scaleZ = modelInstance.transform.getScaleZ();
        float scale = Block.getBlock(bodyID).getScale()*0.3f;
        modelInstance.transform.scale(scale,scale,scale);
        Graphic.getModelBatch().render(modelInstance, Graphic.ENVIRONMENT);
        modelInstance.transform.scale(scaleX,scaleY,scaleZ);
    }



    public Container getDrop() {
        return dropData.get();
    }
}

File: core/src/com/intbyte/bw/engine/entity/Drop.kt
package com.intbyte.bw.engine.entity

import com.badlogic.gdx.graphics.g3d.ModelInstance
import com.intbyte.bw.engine.block.Block
import com.intbyte.bw.engine.graphic.Graphic
import com.intbyte.bw.engine.item.Container
import com.intbyte.bw.engine.physic.PhysicData
import com.intbyte.bw.engine.utils.gdx.renderWithTransform

open class Drop(blockID: Int, dropData: DropData) : Entity() {
    private var modelInstance: ModelInstance
    private var dropData: DropData
    override fun spawn() {
        val data = getBody().userData as PhysicData
        data.setObject(this)
        data.type = PhysicData.DROP
    }

    override fun getEntityModel(): ModelInstance {
        return modelInstance
    }

    override fun render() {
        renderWithTransform(0f,0f, 0f, Block.getBlock(bodyID).scale * 0.3f) {
            Graphic.getModelBatch().render(modelInstance, Graphic.getEnvironment(false))
        }
    }

    val drop: Container
        get() = dropData.get()

    init {
        rotate = Math.random().toFloat() * 360
        modelInstance = Block.getBlock(blockID).modelInstance
        this.dropData = dropData
    }
}